File: leetcode/30-day-leetcoding-challenge/30-day-leetcoding-challenge/week-2/day-11-diameter-of-binary-tree/recursion.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        ans = 1
        def helper(node) :
            nonlocal ans
            if not node : return 0
            left = helper(node.left)
            right = helper(node.right)
            ans = max(left + right + 1 , ans)
            return max(left , right) + 1
        helper(root)
        return ans - 1
        

File: leetcode/practice/145-binary-tree-postorder-traversal/recursion.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) ;
    void helper(TreeNode* root , vector<int>& ans) ;
};

void Solution :: helper(TreeNode* root , vector<int>& ans)
{
    if(!root)return ;
    helper(root->left , ans) ;
    helper(root->right , ans) ;
    ans.push_back(root->val) ;
}

vector<int> Solution :: postorderTraversal(TreeNode* root)
{
    vector<int> ans ;
    helper(root , ans) ;
    return ans ;
}



